lib/bundler/cli/list.rb

```diff
@@ -7,19 +7,54 @@ def initialize(options)
     end
 
     def run
+      raise InvalidOption, "The `--only` and `--without` options cannot be used together" if @options[:only] && @options[:without]
+
+      raise InvalidOption, "The `--name-only` and `--paths` options cannot be used together" if @options["name-only"] && @options[:paths]
+
       specs = Bundler.load.specs.reject {|s| s.name == "bundler" }.sort_by(&:name)
 
-      raise InvalidOption, "The `--name-only` and `--paths` options cannot be used together" if @options["name-only"] && @options["paths"]
+      specs = filtered_specs_by_groups(specs) if @options[:only] || @options[:without]
+
+      return Bundler.ui.info "No gems in the Gemfile" if specs.empty?
+
       return specs.each {|s| Bundler.ui.info s.name } if @options["name-only"]
       return specs.each {|s| Bundler.ui.info s.full_gem_path } if @options["paths"]
 
-      return Bundler.ui.info "No gems in the Gemfile" if specs.empty?
       Bundler.ui.info "Gems included by the bundle:"
-      specs.each do |s|
-        Bundler.ui.info "  * #{s.name} (#{s.version}#{s.git_version})"
-      end
+
+      specs.each {|s| Bundler.ui.info "  * #{s.name} (#{s.version}#{s.git_version})" }
 
       Bundler.ui.info "Use `bundle info` to print more detailed information about a gem"
     end
+
+  private
+
+    def verify_group_exists
+      raise InvalidOption, "`#{@options[:without]}` group could not be found." if @options[:without] && !@definition.groups.include?(@options[:without].to_sym)
+
+      raise InvalidOption, "`#{@options[:only]}` group could not be found." if @options[:only] && !@definition.groups.include?(@options[:only].to_sym)
+    end
+
+    def filtered_specs_by_groups(specs)
+      builder = Dsl.new
```

Comment:
Why are you parsing the Gemfile again?

Reply:
To filter by groups. Does spec set contains gems groups?

Reply:
the definition has the ability to restrict to a group (and its transitive dependencies). See `Definition#specs_for`

Reply:
Also added conditional for groups case